File: src/CircleAI.Speech.Cloud/AssemblyAiSpeechRecognizer.cs
// AssemblyAiSpeechRecognizer.cs
//
// (3.3.0) ISpeechRecognizer backed by AssemblyAI. Two-step flow:
// upload bytes → POST /v2/transcript with upload_url → poll until
// status=completed. Bytes are wrapped as WAV for compatibility.

using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace CircleAI.Speech.Cloud;

/// <summary>(3.3.0) AssemblyAI-backed <see cref="ISpeechRecognizer"/>.</summary>
public sealed class AssemblyAiSpeechRecognizer : ISpeechRecognizer
{
    private readonly HttpClient _http;
    private readonly AssemblyAiOptions _options;
    private readonly ILogger _logger;

    public AssemblyAiSpeechRecognizer(HttpClient http, AssemblyAiOptions options, ILogger<AssemblyAiSpeechRecognizer>? logger = null)
    {
        _http    = http    ?? throw new ArgumentNullException(nameof(http));
        _options = options ?? throw new ArgumentNullException(nameof(options));
        _logger  = (ILogger?)logger ?? NullLogger.Instance;

        if (_http.BaseAddress is null) _http.BaseAddress = options.BaseAddress;
    }

    public string BackendId    => "assemblyai";
    public bool   IsConfigured => !string.IsNullOrWhiteSpace(_options.ApiKey);

    public async ValueTask<TranscriptionResult> TranscribeAsync(
        ReadOnlyMemory<byte> audioPcm16Mono,
        int                  sampleRateHz,
        string?              languageHint = null,
        CancellationToken    ct           = default)
    {
        if (!IsConfigured) return Empty();

        // 1) Upload audio.
        var wav = WrapPcmAsWav(audioPcm16Mono, sampleRateHz);
        using var uploadMsg = new HttpRequestMessage(HttpMethod.Post, "/v2/upload")
        {
            Content = new ByteArrayContent(wav),
        };
        uploadMsg.Content!.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");
        uploadMsg.Headers.Add("Authorization", _options.ApiKey);

        using var uploadResp = await _http.SendAsync(uploadMsg, ct).ConfigureAwait(false);
        if (!uploadResp.IsSuccessStatusCode)
        {
            _logger.LogWarning("AssemblyAI upload returned {Status}", uploadResp.StatusCode);
            return Empty();
        }
        using var uploadDoc = await JsonDocument.ParseAsync(
            await uploadResp.Content.ReadAsStreamAsync(ct).ConfigureAwait(false),
            cancellationToken: ct).ConfigureAwait(false);

        var uploadUrl = uploadDoc.RootElement.TryGetProperty("upload_url", out var u) ? u.GetString() : null;
        if (string.IsNullOrWhiteSpace(uploadUrl)) return Empty();

        // 2) Submit transcript job.
        var body = new StringBuilder("{");
        body.Append($"\"audio_url\":\"{uploadUrl}\",");
        body.Append($"\"speech_model\":\"{_options.SpeechModel}\"");
        if (!string.IsNullOrWhiteSpace(languageHint))
        {
            body.Append($",\"language_code\":\"{languageHint}\"");
        }
        body.Append('}');

        using var submitMsg = new HttpRequestMessage(HttpMethod.Post, "/v2/transcript")
        {
            Content = new StringContent(body.ToString(), Encoding.UTF8, "application/json"),
        };
        submitMsg.Headers.Add("Authorization", _options.ApiKey);

        using var submitResp = await _http.SendAsync(submitMsg, ct).ConfigureAwait(false);
        if (!submitResp.IsSuccessStatusCode)
        {
            _logger.LogWarning("AssemblyAI submit returned {Status}", submitResp.StatusCode);
            return Empty();
        }
        using var submitDoc = await JsonDocument.ParseAsync(
            await submitResp.Content.ReadAsStreamAsync(ct).ConfigureAwait(false),
            cancellationToken: ct).ConfigureAwait(false);

        var transcriptId = submitDoc.RootElement.TryGetProperty("id", out var id) ? id.GetString() : null;
        if (string.IsNullOrWhiteSpace(transcriptId)) return Empty();

        // 3) Poll until completed (max 60 attempts of 500 ms = 30 s).
        for (int attempt = 0; attempt < 60; attempt++)
        {
            ct.ThrowIfCancellationRequested();
            await Task.Delay(500, ct).ConfigureAwait(false);

            using var pollMsg = new HttpRequestMessage(HttpMethod.Get, $"/v2/transcript/{transcriptId}");
            pollMsg.Headers.Add("Authorization", _options.ApiKey);

            using var pollResp = await _http.SendAsync(pollMsg, ct).ConfigureAwait(false);
            if (!pollResp.IsSuccessStatusCode) continue;

            using var pollDoc = await JsonDocument.ParseAsync(
                await pollResp.Content.ReadAsStreamAsync(ct).ConfigureAwait(false),
                cancellationToken: ct).ConfigureAwait(false);

            var status = pollDoc.RootElement.TryGetProperty("status", out var s) ? s.GetString() : null;
            if (status == "completed")
            {
                var text = pollDoc.RootElement.TryGetProperty("text", out var t) ? t.GetString() ?? "" : "";
                var lang = pollDoc.RootElement.TryGetProperty("language_code", out var lc) ? lc.GetString() : languageHint;
                var duration = pollDoc.RootElement.TryGetProperty("audio_duration", out var ad) && ad.ValueKind == JsonValueKind.Number
                    ? TimeSpan.FromSeconds(ad.GetDouble())
                    : TimeSpan.Zero;

                var segments = new List<TranscribedSegment>();
                if (pollDoc.RootElement.TryGetProperty("words", out var words) && words.ValueKind == JsonValueKind.Array)
                {
                    foreach (var w in words.EnumerateArray())
                    {
                        var start = w.TryGetProperty("start", out var ws) ? ws.GetDouble() / 1000d : 0d;
                        var end   = w.TryGetProperty("end",   out var we) ? we.GetDouble() / 1000d : start;
                        segments.Add(new TranscribedSegment(
                            Text:       w.TryGetProperty("text",       out var wt) ? wt.GetString() ?? "" : "",
                            Offset:     TimeSpan.FromSeconds(start),
                            Duration:   TimeSpan.FromSeconds(Math.Max(0, end - start)),
                            Language:   lang,
                            Confidence: w.TryGetProperty("confidence", out var wc) ? (float)wc.GetDouble() : 0f));
                    }
                }

                return new TranscriptionResult(text, lang, segments, duration);
            }
            if (status == "error")
            {
                var err = pollDoc.RootElement.TryGetProperty("error", out var e) ? e.GetString() : null;
                _logger.LogWarning("AssemblyAI transcript error: {Error}", err);
                return Empty();
            }
        }

        _logger.LogWarning("AssemblyAI transcript {Id} timed out after 30 s", transcriptId);
        return Empty();
    }

    private static TranscriptionResult Empty() =>
        new(string.Empty, null, Array.Empty<TranscribedSegment>(), TimeSpan.Zero);

    private static byte[] WrapPcmAsWav(ReadOnlyMemory<byte> pcm, int sampleRate)
    {
        const int channels = 1, bitsPerSample = 16;
        var byteRate = sampleRate * channels * (bitsPerSample / 8);
        var blockAlign = channels * (bitsPerSample / 8);
        var dataSize = pcm.Length;
        var chunkSize = 36 + dataSize;
        var buffer = new byte[44 + dataSize];
        var span = buffer.AsSpan();
        span[0] = (byte)'R'; span[1] = (byte)'I'; span[2] = (byte)'F'; span[3] = (byte)'F';
        BitConverter.GetBytes(chunkSize).CopyTo(span[4..]);
        span[8] = (byte)'W'; span[9] = (byte)'A'; span[10] = (byte)'V'; span[11] = (byte)'E';
        span[12] = (byte)'f'; span[13] = (byte)'m'; span[14] = (byte)'t'; span[15] = (byte)' ';
        BitConverter.GetBytes(16).CopyTo(span[16..]);
        BitConverter.GetBytes((short)1).CopyTo(span[20..]);
        BitConverter.GetBytes((short)channels).CopyTo(span[22..]);
        BitConverter.GetBytes(sampleRate).CopyTo(span[24..]);
        BitConverter.GetBytes(byteRate).CopyTo(span[28..]);
        BitConverter.GetBytes((short)blockAlign).CopyTo(span[32..]);
        BitConverter.GetBytes((short)bitsPerSample).CopyTo(span[34..]);
        span[36] = (byte)'d'; span[37] = (byte)'a'; span[38] = (byte)'t'; span[39] = (byte)'a';
        BitConverter.GetBytes(dataSize).CopyTo(span[40..]);
        pcm.Span.CopyTo(span[44..]);
        return buffer;
    }
}

File: python/src/circle_ai/speech/cloud/assemblyai_speech_recognizer.py
# speech/cloud/assemblyai_speech_recognizer.py
#
# Port of CircleAI.Speech.Cloud/AssemblyAiSpeechRecognizer.cs (C# — the EXACT spec).
#
# (3.3.0) ISpeechRecognizer backed by AssemblyAI. Three-step flow: upload
# WAV-wrapped bytes to /v2/upload -> POST /v2/transcript with the upload_url ->
# poll /v2/transcript/{id} until status=completed (max 60 attempts of 500 ms =
# 30 s). Auth is the bare api key in the Authorization header. Fail-soft on
# missing key / non-2xx / timeout / error.
#
# The C# drives HttpClient directly (ByteArrayContent upload / StringContent JSON
# submit / GET poll, Task.Delay between polls). The Python port injects the shared
# circle_ai.integration.http.IHttpFetcher: the upload body rides body_bytes with
# application/octet-stream, the submit body rides body_json, and asyncio.sleep
# reproduces the poll delay. JSON is read off resp.json(). Word offsets are in ms
# (start/end / 1000) and audio_duration in seconds.

from __future__ import annotations

import asyncio
import logging
from datetime import timedelta
from typing import List, Optional

from ...integration.http import HttpRequest, IHttpFetcher
from ..contracts import ISpeechRecognizer, TranscribedSegment, TranscriptionResult
from ._audio_http import combine_uri, is_null_or_whitespace, wrap_pcm_as_wav
from .options import AssemblyAiOptions

_logger = logging.getLogger("CircleAI.Speech.Cloud.AssemblyAiSpeechRecognizer")


def _empty() -> TranscriptionResult:
    return TranscriptionResult("", None, (), timedelta(0))


class AssemblyAiSpeechRecognizer(ISpeechRecognizer):
    """(3.3.0) AssemblyAI-backed :class:`ISpeechRecognizer`.

    Mirrors ``CircleAI.Speech.Cloud.AssemblyAiSpeechRecognizer``.
    """

    def __init__(
        self,
        http: IHttpFetcher,
        options: AssemblyAiOptions,
        logger: Optional[logging.Logger] = None,
    ) -> None:
        if http is None:
            raise ValueError("http must not be None")
        if options is None:
            raise ValueError("options must not be None")
        self._http = http
        self._options = options
        self._logger = logger if logger is not None else _logger

    @property
    def backend_id(self) -> str:
        return "assemblyai"

    @property
    def is_configured(self) -> bool:
        return not is_null_or_whitespace(self._options.api_key)

    def _url(self, path: str) -> str:
        return combine_uri(self._options.base_address, path)

    def _auth(self) -> dict:
        return {"Authorization": self._options.api_key or ""}

    async def transcribe_async(
        self,
        audio_pcm16_mono: bytes,
        sample_rate_hz: int,
        language_hint: Optional[str] = None,
        ct: object = None,
    ) -> TranscriptionResult:
        if not self.is_configured:
            return _empty()

        # 1) Upload audio.
        wav = wrap_pcm_as_wav(audio_pcm16_mono, sample_rate_hz)
        upload_resp = await self._http.send_async(
            HttpRequest(
                method="POST",
                url=self._url("/v2/upload"),
                headers=self._auth(),
                body_bytes=wav,
                content_type="application/octet-stream",
            )
        )
        if not upload_resp.is_success:
            self._logger.warning("AssemblyAI upload returned %s", upload_resp.status_code)
            return _empty()
        upload_doc = upload_resp.json()
        upload_url = (
            upload_doc.get("upload_url")
            if isinstance(upload_doc, dict) and isinstance(upload_doc.get("upload_url"), str)
            else None
        )
        if is_null_or_whitespace(upload_url):
            return _empty()

        # 2) Submit transcript job.
        submit_body = {"audio_url": upload_url, "speech_model": self._options.speech_model}
        if not is_null_or_whitespace(language_hint):
            submit_body["language_code"] = language_hint

        submit_resp = await self._http.send_async(
            HttpRequest(
                method="POST",
                url=self._url("/v2/transcript"),
                headers=self._auth(),
                body_json=submit_body,
            )
        )
        if not submit_resp.is_success:
            self._logger.warning("AssemblyAI submit returned %s", submit_resp.status_code)
            return _empty()
        submit_doc = submit_resp.json()
        transcript_id = (
            submit_doc.get("id")
            if isinstance(submit_doc, dict) and isinstance(submit_doc.get("id"), str)
            else None
        )
        if is_null_or_whitespace(transcript_id):
            return _empty()

        # 3) Poll until completed (max 60 attempts of 500 ms = 30 s).
        for _attempt in range(60):
            await asyncio.sleep(0.5)

            poll_resp = await self._http.send_async(
                HttpRequest(
                    method="GET",
                    url=self._url(f"/v2/transcript/{transcript_id}"),
                    headers=self._auth(),
                )
            )
            if not poll_resp.is_success:
                continue

            poll_doc = poll_resp.json()
            if not isinstance(poll_doc, dict):
                continue

            status = poll_doc.get("status") if isinstance(poll_doc.get("status"), str) else None
            if status == "completed":
                text = poll_doc.get("text") if isinstance(poll_doc.get("text"), str) else ""
                lang = (
                    poll_doc.get("language_code")
                    if isinstance(poll_doc.get("language_code"), str)
                    else language_hint
                )
                duration = (
                    timedelta(seconds=float(poll_doc["audio_duration"]))
                    if isinstance(poll_doc.get("audio_duration"), (int, float))
                    else timedelta(0)
                )

                segments: List[TranscribedSegment] = []
                words = poll_doc.get("words")
                if isinstance(words, list):
                    for w in words:
                        if not isinstance(w, dict):
                            continue
                        start = (
                            float(w["start"]) / 1000.0
                            if isinstance(w.get("start"), (int, float))
                            else 0.0
                        )
                        end = (
                            float(w["end"]) / 1000.0
                            if isinstance(w.get("end"), (int, float))
                            else start
                        )
                        confidence = (
                            float(w["confidence"])
                            if isinstance(w.get("confidence"), (int, float))
                            else 0.0
                        )
                        segments.append(
                            TranscribedSegment(
                                text=w.get("text") if isinstance(w.get("text"), str) else "",
                                offset=timedelta(seconds=start),
                                duration=timedelta(seconds=max(0.0, end - start)),
                                language=lang,
                                confidence=confidence,
                            )
                        )

                return TranscriptionResult(text, lang, tuple(segments), duration)
            if status == "error":
                err = poll_doc.get("error") if isinstance(poll_doc.get("error"), str) else None
                self._logger.warning("AssemblyAI transcript error: %s", err)
                return _empty()

        self._logger.warning("AssemblyAI transcript %s timed out after 30 s", transcript_id)
        return _empty()


__all__ = ["AssemblyAiSpeechRecognizer"]
